File: kt03/src/main/java/com/btime/kt03/instance/java/LazyThreadSafeDoubleCheck.java
package com.btime.kt03.instance.java;

public class LazyThreadSafeDoubleCheck {
    private static volatile LazyThreadSafeDoubleCheck INSTANCE;

    private LazyThreadSafeDoubleCheck(){}

    public static LazyThreadSafeDoubleCheck getInstance(){
        if(INSTANCE == null){
            synchronized (LazyThreadSafeDoubleCheck.class){
                if(INSTANCE == null){
                    //初始化分为实例化和赋值两步，尽管我们把这一步写成下面的语句，
                    //但Java虚拟机并不保证其他线程这两步的顺序究竟是怎么样的
                    INSTANCE = new LazyThreadSafeDoubleCheck();
                }
            }
        }
        return INSTANCE;
    }
}

File: kt03/src/main/java/com/btime/kt03/instance/kotlin/LazyThreadSafeDoubleCheck.kt
package com.btime.kt03.instance.kotlin

class LazyThreadSafeDoubleCheck private constructor(){
    companion object{
        private val instance by lazy(mode = LazyThreadSafetyMode.SYNCHRONIZED){
            LazyThreadSafeDoubleCheck()
        }

        private @Volatile var instance2: LazyThreadSafeDoubleCheck ?= null

        fun get():LazyThreadSafeDoubleCheck{
            if(instance2 == null){
                synchronized(this){
                    if(instance2 == null){
                        instance2 = LazyThreadSafeDoubleCheck()
                    }
                }
            }
            return instance2!!
        }
    }
}